### PR TITLE
WIP: support github app repo cloning and authorization (#3086, #4348)

### LINKED ISSUE #3086: Support private GitHub repo authentication using GitHub app
# Summary

Support private repositories authentication using [GitHub app authentication](
https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/#http-based-git-access-by-an-installation).
# Motivation

We want to use argocd with multiple private repositories. The current options are:

1. Create a deploy key for each repository and upload them to argocd (hard to manage)
2. Create a user for argocd (expensive, as you need to pay for a seat in the organization).

# Proposal

Add support for GitHub app authentication. Allow the user to define the credentials using something like: 
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: argocd-cm
  namespace: argocd
  labels:
    app.kubernetes.io/name: argocd-cm
    app.kubernetes.io/part-of: argocd
data:
  repositories: |
    - url: https://github.com/argoproj/private-repo
    - url: https://github.com/argoproj/other-private-repo
  repository.credentials: |
    - url: https://github.com/argoproj
      githubAppPrivateKey:
        name: my-secret
        key: privateKey
      githubInstallationId: <>
```

### LINKED ISSUE #4348: Support GitHub organization app for git cloning.
# Summary

Instead of using username, password/token or ssh deploy keys.
It would be nice to support GitHub apps.

Much like
https://github.com/jenkinsci/github-branch-source-plugin/blob/master/docs/github-app.adoc

# Motivation

Reduce complexity in have multiple repos argocd can access across an GitHub organization

# Proposal

Could be implemented using https://github.com/bradleyfalzon/ghinstallation

### PR DESCRIPTION
**Pull request by jetersen**
fixes #4348 
fixes #3086 

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)).

### CODE REVIEW

File: util/git/creds.go [hunk 1]
```diff
@@ -179,3 +182,103 @@ func (c SSHCreds) Environ() (io.Closer, []string, error) {
...
+	authorization := fmt.Sprintf("Bearer %s", bearer)
+	accept := "application/vnd.github.v3+json"
+
+	url := fmt.Sprintf("%s/repos/%s/%s/installation", baseURL, owner, repo)
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	req.Header.Add("Authorization", authorization)
```
Comment (jetersen) [inline util/git/creds.go:237]: Hmm, properly better off using a custom transport for these headers.

File: util/git/creds.go [hunk 2]
```diff
@@ -179,3 +182,103 @@ func (c SSHCreds) Environ() (io.Closer, []string, error) {
...
+	if err != nil {
+		return nil, nil, err
+	}
+
+	defer resp.Body.Close()
+
+	// TODO handle response code
+
+	var access github.InstallationAccessToken
+	json.NewDecoder(resp.body).decode(&access)
```
Comment (jetersen) [inline util/git/creds.go:274]: This should be moved to github package to remove the complexity here.

**Comment by diranged** [comment]
Is this code going to move anywhere? I would love to see the feature get added..

**Comment by jetersen** [comment]
fixed by https://github.com/argoproj/argo-cd/pull/5355

**Pull request closed without merging**